### ISSUE TITLE
Using Poetry for Dependency Management & Packaging? [DISCUSSION]

### ISSUE THREAD

**Issue by AlexAndorra** [pymc-devs/pymc#3769]
Hi everyone!
As discussed with @eigenfoo, here is an issue to discuss whether it would be appropriate to use [Poetry](https://python-poetry.org/) for dependency management and packaging.

This would avoid having two requirements files and automate requirements' versions management. I use it in other projects and I find it better than traditional tools like pip and conda. Happy to exchange about it and possibly implement it!

**Comment by rpgoldman** [comment]
@AlexAndorra would you see if you can enumerate the advantages of Poetry? Your message doesn't really say in what ways you think poetry is better.  Any time we change foundational software, we cost ourselves unproductive time in learning new tools that is not time spent on our core activities.  Sometimes this cost pays off in the long term in greater efficiency, etc.  But sometimes not.  So I would like to take a *very* hard look at this before we jump.

Two personal notes:

1. I had poetry added to another project, by a team-mate, and I'm not at all happy about it.  In addition to being one more bit of infrastructure to learn -- everyone who uses Python *must* understand pip at least a little, so pip is "free" --  and I don't understand how poetry plays when I develop on projects with multiple different packages being integrated.

2. What would the impact be on those of us who use virtual environments instead of Conda?  I exclusively use virtualenvwrapper, not even Conda, and it's been adequate to my needs.

**Comment by AlexAndorra** [comment]
Hey @rpgoldman ! First, I don't have a strong prior on using Poetry here -- this really is a discussion to see if it would be adequate, rather than me advocating for one solution or the other. To be precise, I'd say my prior is 60% in favor of Poetry. To answer your questions and explain a little more why I thought this could be a useful tool for PyMC:

- **Deterministic build, managed automatically**: On my website, I developped on a conda env and often found myself with version issues when I deployed on the server (the setup.py was not deterministic enough): e.g it turned out I developped on pandas 0.25 while the server had installed pandas 0.24. In the end, I had to clarify the exact version of packages in the setup.py manually to make sure my code wouldn't break on the server, which was cumbersome, time-consuming and boring. 
Poetry handles that automatically: packages' versions that are compatible with each other are written automatically in the lock file. As a result, there is no dependency conflicts, I don't need one requirements file for prod and another for dev, and my build really is deterministic: the environment I build on is the same I deploy on. Bonus: It's easy to check the state of your project's dependencies.

- I don't know how you manage building and deployment of PyMC to PyPI, so I don't know your pain points. But Poetry makes it easy to **build source distribution and wheels and then publish on PyPI** from the command line.

- **Impact for those using virtual envs**: I use conda or Poetry so I'm no expert here, but I know Poetry either uses your configured virtualenvs or automatically creates one when installing a new project.

- I understand your point about losing time in the short run to gain more time in the long run. Obviously I didn't have the same experience as you did, as I found Poetry's CLI easy and intuitive, but that's eminently subjective!

For a more general view, [the documentation](https://python-poetry.org/docs/) is really good, and I also advise [this short article](https://jacobian.org/2019/nov/11/python-environment-2020/) by the co-creator of Django: he explains his requirements, his setup and in particular why he chose Poetry.

I hope all of that is useful and answers your questions :) I'm really curious to discuss it with all of you!
PyMCheers 🖖
